sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala

```diff
@@ -0,0 +1,237 @@
+  extends BinaryExpression with ImplicitCastInputTypes with NullIntolerant with Serializable {
+  private def format: Expression = right
+
+  private lazy val formatString = if (format.foldable) {
+    format.eval().toString.toUpperCase(Locale.ROOT)
+  } else {
+    ""
+  }
+  private lazy val parser = new MaskDigitSequenceParser(formatString, nullOnError)
+
+  override def prettyName: String = functionName
+  override def dataType: DataType = StringType
+  override def inputTypes: Seq[DataType] = Seq(StringType, StringType)
+  override def checkInputDataTypes(): TypeCheckResult = {
+    val inputTypeCheck = super.checkInputDataTypes()
+    if (inputTypeCheck.isSuccess) {
+      val formatStringValid = formatString.forall {
+        case 'x' | 'X' | '-' => true
+        case ch if ch.isDigit || ch.isWhitespace => true
+        case _ => false
+      }
+      if (formatStringValid) {
+        TypeCheckResult.TypeCheckSuccess
+      } else {
+        throw QueryCompilationErrors.maskCcnInvalidFormatError(prettyName)
+      }
+    } else {
+      inputTypeCheck
+    }
+  }
+
+  override def nullSafeEval(string: Any, format: Any): Any =
+    parser.parse(string.asInstanceOf[UTF8String])
+
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    val builder = ctx.addReferenceObj(
+      "parser", parser, classOf[MaskDigitSequenceParser].getName)
+    val eval = left.genCode(ctx)
+    ev.copy(code =
+      code"""
+        |${eval.code}
+        |boolean ${ev.isNull} = ${eval.isNull};
+        |${CodeGenerator.javaType(dataType)} ${ev.value} = ${CodeGenerator.defaultValue(dataType)};
+        |if (!${ev.isNull}) {
+        |  ${ev.value} = $builder.parse(${eval.value});
+        |  if (${ev.value} == null) {
+        |    ${ev.isNull} = true;
+        |  }
+        |}
+      """.stripMargin)
+  }
+}
+
+/** Executes the string parsing steps for the MaskDigitSequence class. */
+class MaskDigitSequenceParser(formatString: String, nullOnError: Boolean) extends Serializable {
+  def parse(input: UTF8String): UTF8String = {
+    val inputString = input.toString
+    var formatStringIndex = 0
+    var error = false
+    def skipFormatWhitespace(): Unit =
+      while (formatStringIndex < formatString.length &&
+        formatString(formatStringIndex).isWhitespace) {
+        formatStringIndex += 1
+      }
+    // Check and consume each character in the input string, comparing against characters in the
+    // format string.
+    val result = inputString.map { inputChar =>
+      if (error) {
+        // If we have encountered an error, leave the input character alone; we will raise an
+        // exception or return NULL after this loop has finished.
+        inputChar
+      } else if (inputChar.isWhitespace) {
+        // The input character is whitespace. Ignore it and continue comparing the next input
+        // character against the same format string character as this iteration.
+        inputChar
```

Comment:
Shouldn't we check if the corresponding format string char is space as well?

Reply:
Good question, I think the intention is to skip and ignore whitespace in both the input string and the format string. For example, for use cases like credit card numbers, phone numbers, and social security numbers, these are not material to the data in the field. I put a comment to this effect here.

Reply:
Update on this: per our offline conversation, we decided to require whitespace in the format string to match the input string exactly. I updated the PR accordingly.